adodb-lib.inc.php

```diff
@@ -1208,64 +1249,156 @@ function _adodb_debug_execute($zthis, $sql, $inputarr)
-				else ADOConnection::outp("-----<hr>\n($dbt): ".$sqlTxt."$ss\n-----<hr>\n",false);
-
+		{
+			if ($inBrowser)
+			{
+				$outString = "<hr class='adodb-debug'>(%s):  %s &nbsp; %s<hr class='adodb-debug'>";
+				ADOConnection::outp(sprintf($outString, $myDatabaseType,htmlspecialchars($sqlTxt),$ss),false);
+			}
+			else
+			{				
+				$outString = sprintf("%s\n%s\n    %s\n%s\n",str_repeat('-',78),$myDatabaseType,$sqlTxt,$ss,str_repeat('-',78));
+				ADOConnection::outp($outString,false);
+			}
+		}
+		
+		/*
+		* Send last error to output
+		*/
 		ADOConnection::outp($zthis->ErrorNo() .': '. $zthis->ErrorMsg());
 	}
 
-	if ($zthis->debug === 99) _adodb_backtrace(true,9999,2);
+	if ($zthis->debug === 99) 
+	{
+		_adodb_backtrace(true,9999,2);
+	}
 	return $qID;
 }
 
-# pretty print the debug_backtrace function
-function _adodb_backtrace($printOrArr=true,$levels=9999,$skippy=0,$ishtml=null)
+/**
+ * pretty print the debug_backtrace function
+ *
+ * @param string[]|bool $printOrArr			Whether to print the result directly or return the result
+ * @param int			$maximumDepth		The maximum depth of the array to traverse
+ * @param int			$elementsToIgnore	The backtrace array indexes to ignore
+ * @param null|bool  	$ishtml			  	Are we in a CLI or CGI environment
+ *
+ * @return void
+ */
+function _adodb_backtrace($printOrArr=true,$maximumDepth=9999,$elementsToIgnore=0,$ishtml=null)
 {
-	if (!function_exists('debug_backtrace')) return '';
+	if (!function_exists('debug_backtrace')) 
+		return '';
 
-	if ($ishtml === null) $html = (isset($_SERVER['HTTP_USER_AGENT']));
-	else $html = $ishtml;
+	if ($ishtml === null)
+		/*
+		Auto determine if we in a CGI enviroment
+		*/
+		$html = (isset($_SERVER['HTTP_USER_AGENT']));
+	else 
+		$html = $ishtml;
 
-	$fmt = ($html) ? "</font><font color=#808080 size=-1> %% line %4d, file: <a href=\"file:/%s\">%s</a></font>" : "%% line %4d, file: %s";
+	$cgiString = "</font><font color=#808080 size=-1> %% line %4d, file: <a href=\"file:/%s\">%s</a></font>";
+	$cliString =  "%% line %4d, file: %s";
+	$fmt = ($html) ? $cgiString : $cliString;
 
 	$MAXSTRLEN = 128;
 
 	$s = ($html) ? '<pre align=left>' : '';
+	
+	
+	if (is_array($printOrArr)) 
+		$traceArr = $printOrArr;
+	else 
+		$traceArr = debug_backtrace();
+	
 
-	if (is_array($printOrArr)) $traceArr = $printOrArr;
-	else $traceArr = debug_backtrace();
+	/*
+	* Remove first 2 elements that just show calls to  adodb_backtrace
+	*/
 	array_shift($traceArr);
 	array_shift($traceArr);
-	$tabs = sizeof($traceArr)-2;
 
-	foreach ($traceArr as $arr) {
-		if ($skippy) {$skippy -= 1; continue;}
-		$levels -= 1;
-		if ($levels < 0) break;
+	/*
+	* We want last element to have no indent
+	*/
+	$tabs = sizeof($traceArr) - 1;
+	
+	
+	foreach ($traceArr as $arr) 
+	{
+		if ($elementsToIgnore) 
+		{
+			/*
+			* Ignore array element at start of array
+			*/
+			$elementsToIgnore--; 
```

Comment:
Add `$tabs--;` here, otherwise we over-indent the trace when $elementsToIgnore > 0